plugins/newsdownloader.koplugin/feed_source.lua

```diff
@@ -0,0 +1,398 @@
+local BD = require("ui/bidi")
+local DownloadBackend = require("epubdownloadbackend")
+local NewsHelpers = require("http_utilities")
+local dateparser = require("lib.dateparser")
+local logger = require("logger")
+local md5 = require("ffi/sha2").md5
+local util = require("util")
+local _ = require("gettext")
+local FFIUtil = require("ffi/util")
+local T = FFIUtil.template
+
+local FeedSource = {
+    file_extension = ".epub"
+}
+
+function FeedSource:new(o)
+    o = o or {}
+    self.__index = self
+    setmetatable(o, self)
+    return o
+end
+
+function FeedSource:getInitializedFeeds(feed_list, progress_callback, error_callback)
+    local initialized_feeds = {}
+    local unsupported_feeds_urls = {}
+
+    for idx, feed in ipairs(feed_list) do
+        local url = feed[1]
+        -- Show a UI update
+        progress_callback(T(
+            _("Setting up feed %1 of %2."),
+            idx,
+            url
+        ))
+        -- Initialize the feed
+        local ok, response = pcall(function()
+            return self:initializeDocument(
+                self:fetchDocumentByUrl(url)
+            )
+        end)
+        -- If the initialization worked, add the feed
+        -- to a list of initialized feeds
+        if ok and response then
+            table.insert(initialized_feeds, {
+                config = feed,
+                document = response,
+            })
+        else
+            logger.dbg("FeedSource: Unsupported feed ", response)
+            table.insert(unsupported_feeds_urls, {
+                url .. ": " .. response
+            })
+        end
+    end
+
+    if #unsupported_feeds_urls > 0 then
+        -- When some errors are present, we get a sour message that includes
+        -- information about the source of the error.
+        local unsupported_urls = ""
+        for key, value in pairs(unsupported_feeds_urls) do
+            -- Create the error message.
+            --            unsupported_urls = unsupported_urls .. " " .. value[1] .. " " .. value[2]
+            unsupported_urls = value[1] .. "\n\n"
+            -- Not sure what this does.
+            if key ~= #unsupported_feeds_urls then
+                unsupported_urls = BD.url(unsupported_urls) .. ", "
+            end
+        end
+        error_callback(
+            T(_("Could not initialize a feed:\n\n%2\n\nPlease review your feed configuration.", "Could not initialize %1 feeds:\n\n%2\n\nPlease review your feed configurations.", #unsupported_urls),
+                #unsupported_urls, unsupported_urls)
```

Comment:
Hold on, you can't have tested this. ;-) My suggestion was conceptual, not final.

It should be closer to this:
```suggestion
            T(N_("Could not initialize a feed:\n\n%2\n\nPlease review your feed configuration.", "Could not initialize %1 feeds:\n\n%2\n\nPlease review your feed configurations.", #unsupported_feeds_urls),
                #unsupported_feeds_urls, unsupported_urls)
```

I didn't bother trying to find the exact variables because you know the code better than I do! ;-) I guess I forgot to make the `N_()` thing sufficiently explicit though.

Reply:
I did test it, but I'd never seen `N_()` so I wasn't sure what to expect from that method. It returns one or the other argument based on a condition, is what I'm guessing it does?

Reply:
Ok, so there is something buggy happening. When a feed cannot be initialized because the URL is incorrect, an error is returned from `self:fetchDocumentByUrl`.

Oddly, the value that is returned becomes: 
```sh
DEBUG plugins/newsdownloader.koplugin/feed_source.lua:103: (Reason: Failed to download feed document)
```
Any idea why this would be happening?
``` lua
function FeedSource:fetchDocumentByUrl(url)
    local document
    -- Get the XML document representing the feed
    local ok, response = pcall(function()
            local success, content = NewsHelpers:getUrlContent(url)
            if (success) then
                return content
            else
                error("Failed to download content for url:", url)
            end
    end)
    -- Check to see if a response is available to deserialize.
    if ok then
        -- Deserialize the XML document into something Lua can use
        document = NewsHelpers:deserializeXMLString(response)
    end
    -- Return the document or any errors that may have occured
    if ok or document then
        return document
    else
        if not ok then
            error("(Reason: Failed to download feed document)")
        else
            error("(Reason: Error during feed document deserialization)")
        end
    end
end
```

Reply:
> It returns one or the other argument based on a condition, is what I'm guessing it does?

Just one or the other is for English. There can be many different plurals. It's there so we don't just get the correct result in English and sufficiently similar languages like Dutch, but also in Russian.

https://github.com/koreader/koreader/blob/ec280f874f0cfce660bdb4d396589939503d3c8b/spec/unit/gettext_spec.lua#L22-L39
https://github.com/koreader/koreader/blob/ec280f874f0cfce660bdb4d396589939503d3c8b/spec/unit/gettext_spec.lua#L79-L88

Reply:
> When an incorrect URL is supplied to self:fetchDocumentByUrl(), the error message returned to the pcall is
> 
> plugins/newsdownloader.koplugin/feed_source.lua:103: (Reason: Failed to download feed document)
> 
> What I expect the message to be is:
> 
> Reason: Failed to download feed document

Did you mess something up during copy/paste? Those are identical?

Reply:
NVM. I didn't realize that [`error()` has an argument to set the error lever](https://www.lua.org/pil/8.5.html).